### ISSUE TITLE
memory_efficient_attention differences with MultiHeadAttention

### ISSUE THREAD

**Issue by pabloppp** [facebookresearch/xformers#687]
Hello. I'm trying to use  `memory_efficient_attention` inside of my model. The model is composed of two different Transformer Encoders, the first one gets a sequence, processes it bidirectionally (no causal mask), and returns an embedding. The second gets another sequence, the embedding gets pre-pended to the sequence and then it is processed causally (next-token prediction).

I'm getting good results using pytorch's MultyHeadAttention, but wanted to try speeding up the training, so I implemented this module to be able to use as a drop-in replacement:
```python
class MultiheadAttentionX(nn.Module):
    def __init__(self, embed_dim, num_heads, dropout=0., bias=True, batch_first=False, causal=False):
        super().__init__()
        self.dropout = dropout
        self.causal = causal
        self.batch_first = batch_first
        self.num_heads = num_heads
        self.in_proj_weight = nn.Parameter(torch.randn(embed_dim*3, embed_dim))
        self.in_proj_bias = nn.Parameter(torch.randn(embed_dim*3)) if bias else None
        self.out_proj = nn.Linear(embed_dim, embed_dim, bias=bias)
        
        torch.nn.init.xavier_uniform_(self.in_proj_weight)
        if self.in_proj_bias is not None:
            nn.init.constant_(self.in_proj_bias, 0)
            nn.init.constant_(self.out_proj.bias, 0)
            
    def forward(self, query, key, value):
        w = self.in_proj_weight.chunk(3, dim=0)
        b = self.in_proj_bias.chunk(3, dim=0)
        if not self.batch_first:
            query, key, value = query.permute(0, 1), key.permute(0, 1), value.permute(0, 1)

        query = nn.functional.linear(query, w[0], bias=b[0]).view(*query.shape[:2], self.num_heads, -1)
        key = nn.functional.linear(key, w[1], bias=b[1]).view(*key.shape[:2], self.num_heads, -1) 
        value = nn.functional.linear(value, w[2], bias=b[2]).view(*value.shape[:2], self.num_heads, -1)
        
        x = xops.memory_efficient_attention(query, key, value, p=self.dropout, attn_bias=xops.LowerTriangularMask() if self.causal else None)
        x = x.reshape(x.size(0), x.size(1), -1)

        x = self.out_proj(x)
        if not self.batch_first:
            x = x.permute(0, 1)

        return (x,)
```
With the same params so I could even load an existing state_dict and switch between the different approaches.
When I try this module instead of MultyHeadAttention, with the same initialized weights, I get a very similar, but not identical output (9e-07 max difference in fp32, and around 5e-4 in pf16), and once I start training, the results diverge a lot.
Here is a comparison of two runs (using AMP) where the only difference is changing MultiHeadAttention with MultiHeadAttentionX:

<img width="941" alt="Captura de pantalla 2023-03-10 a las 0 06 24" src="https://user-images.githubusercontent.com/4237014/224180694-478766ac-e1bc-414a-a0a6-ede1e0559bb6.png">

As you can see, the MultiHeadAttentionX model performs considerably worse. I even considered that maybe the issue could be the initialization, so I started a training using MultiHeadAttention, then after a few iterations, and since the weights are compatible, I just switched to MultiHeadAttentionX, and this happened:

<img width="951" alt="Captura de pantalla 2023-03-10 a las 0 08 45" src="https://user-images.githubusercontent.com/4237014/224181026-da77533b-51e2-49d7-9d37-d50ae560c183.png">

Immediately after switching to MultiHeadAttentionX the loss increased and accuracy dropped.

Is this normal? Is there an issue with my implementation? What is the cause of the difference? I though ops.memory_efficient_attention was "exact attention, not an approximation" but unless I made a mistake in my implementation, this doesn't seem to be the case :( and I'd really love to be able to benefit from the speedup and memory reduction that I can get from using it.

Thank you.

**Comment by danthe3rd** [comment]
>  I get a very similar, but not identical output (9e-07 max difference in fp32, and around 5e-4 in pf16)

This is expected and normal, as the calculation is done a bit differently. It might change the model output because of that, but it shouldn't change the loss/accuracy significantly, so your results are very surprising.

A few questions:
(1) are you using dropout? Did you make sure it's disabled during evaluation? (XFormers won't do that for you)
(2) The idea of loading a state dict is quite interesting, and I would investigate that further. Can you compare the result of each layer and see where it diverges?

**Comment by danthe3rd** [comment]
I'm not sure but it looks like the problem is not from xformers. There is also this parameter that I don't understand (`batch_first=False`) - not sure why the batch dimension wouldn't be the first one.
If I were you, I would try to debug it starting with the first encoder layer. With the same input, I would expect q/k/v to be exactly the same. You can look at the [pytorch implementation ](https://github.com/pytorch/pytorch/blob/728dfeee486fbd965710ccbb225fd275dd7bd35c/torch/ao/nn/quantizable/modules/activation.py#L15)for MultiHeadAttention as it's python code.

**Comment by pabloppp** [comment]
I reviewed my implementation and turns out I was accidentally making both my encoder and decoder causal when using MultiheadAttentionX, hence the big shift in the metrics. I just started a new run with that fixed, and will report here the results, but I think this issue can be closed.

BTW: after fixing this, I was able to load the state_dict of a model trained using  MultiheadAttention **without** `xformers.memory_efficient_attention` into a model that uses it and got the expected results (small differences in the outputs, but good enough for the results to be the same), and I expect to be able to do the opposite as well (train with memory_efficient_attention then replace it with MultiheadAttention to be able to run it in devices that don't support memory_efficient_attention).

🙇 Thank you for your support!
